hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientPartitionServiceImpl.java

```diff
@@ -0,0 +1,258 @@
+                        + ", new connection " + connection);
+            }
+            return true;
+        }
+        if (partitionStateVersion <= current.partitionSateVersion) {
+            if (logger.isFinestEnabled()) {
+                logFailure(connection, partitionStateVersion, current,
+                        "response state version is old");
+            }
+            return false;
+        }
+        return true;
+    }
+
+    private void logFailure(Connection connection, int partitionStateVersion, PartitionTable current, String cause) {
+
+        logger.finest(" We will not apply the response, since " + cause + " . Response is from " + connection
+                + ". Current connection " + current.connection
+                + " response state version:" + partitionStateVersion
+                + ". Current state version: " + current.partitionSateVersion);
+    }
+
+    private Int2ObjectHashMap<Address> convertToPartitionToAddressMap(Collection<Map.Entry<Address, List<Integer>>> partitions) {
+        Int2ObjectHashMap<Address> newPartitions = new Int2ObjectHashMap<Address>();
+        for (Map.Entry<Address, List<Integer>> entry : partitions) {
+            Address address = entry.getKey();
+            for (Integer partition : entry.getValue()) {
+                newPartitions.put(partition, address);
+            }
+        }
+        return newPartitions;
+    }
+
+    public void reset() {
+        partitionTable.set(new PartitionTable(null, -1, new Int2ObjectHashMap<Address>()));
+    }
+
+    @Override
+    public Address getPartitionOwner(int partitionId) {
+        return getPartitions().get(partitionId);
+    }
+
+    private Int2ObjectHashMap<Address> getPartitions() {
+        return partitionTable.get().partitions;
+    }
+
+    @Override
+    public int getPartitionId(@Nonnull Data key) {
+        final int pc = getPartitionCount();
+        if (pc <= 0) {
+            return 0;
+        }
+        int hash = key.getPartitionHash();
+        return HashUtil.hashToIndex(hash, pc);
+    }
+
+    @Override
+    public int getPartitionId(@Nonnull Object key) {
+        final Data data = client.getSerializationService().toData(key);
+        return getPartitionId(data);
+    }
+
+    @Override
+    public int getPartitionCount() {
+        while (partitionCount == 0 && client.getConnectionManager().isAlive()) {
+            ClientClusterService clusterService = client.getClientClusterService();
+            Collection<Member> memberList = clusterService.getMemberList();
+            if (MemberSelectingCollection.count(memberList, MemberSelectors.DATA_MEMBER_SELECTOR) == 0) {
+                throw new NoDataMemberInClusterException(
+                        "Partitions can't be assigned since all nodes in the cluster are lite members");
+            }
+            try {
+                Thread.sleep(BLOCKING_GET_ONCE_SLEEP_MILLIS);
```

Comment:
Why does partition service and member list service work different on fetching the initial list of members/partitions? For member list service you use `waitForInitialMembershipEvents` and make the client start or reconnect to wait until the list is fetched but for partition list fetching we delay this to an invocation until the `getPartitionCount()` is called. Intuitively, I expect both to work similarly since both are states of the cluster and the client can not work until one or the other is completed.

Reply:
We should consider async start case specifically. On async start, we don't wait for both memberlist and partition table in the beginning. An empty member list can be handled via invocation system. 
An empty partition table(zero partition size) can not be handled in invocation system, because getPartitionCount is called before an invocation is created. That is why, this method needs to wait for the partition table. 
We may change this behaviour by setting partition hash to invocation rather than partition id. But that change is too big and not concern of this pr. Lets address it separately.

Reply:
let's open an issue.